oss_src/unity/python/sframe/cython/cy_pylambda_workers.pyx

```diff
@@ -0,0 +1,473 @@
+            
+            x = pyobject_from_flexible_type(v_in[i])
+            self.output_buffer[i] = self.lambda_function(x)
+
+        process_common_typed_list(lcd.output_values, self.output_buffer, lcd.output_enum_type)
+
+    @cython.boundscheck(False)
+    cdef eval_multiple_rows(self, lambda_call_by_dict_data* lcd):
+
+        cdef dict arg_dict = {}
+        cdef long i, j
+        cdef long n = lcd.input_rows[0].size()
+        cdef long n_keys = lcd.input_keys[0].size()
+
+        self._set_dict_keys(lcd.input_keys)
+
+        if len(self.output_buffer) != n:
+            self.output_buffer = [None]*n
+
+        for i in range(n):
+
+            if lcd.input_keys[0][i].size() != n_keys:
+                raise ValueError("Row %d does not have the correct number of rows (%d, should be %d)"
+                                 % (lcd.input_keys[0][i].size(), n))
+
+            arg_dict = self.arg_dict_base.copy()
+            
+            for j in range(n_keys):
+                arg_dict[self.keys[j]] = pyobject_from_flexible_type(lcd.input_rows[0][i][j])
+
+            self.output_buffer[i] = self.lambda_function(arg_dict)
+
+        process_common_typed_list(lcd.output_values, self.output_buffer, lcd.output_enum_type)
+
+    @cython.boundscheck(False)
+    cdef eval_sframe_rows(self, lambda_call_by_sframe_rows_data* lcd):
+
+        cdef dict arg_dict = {}
+        cdef long i, j
+        cdef long n = lcd.input_rows[0].num_rows()
+        cdef long n_keys = lcd.input_keys[0].size()
+
+        assert lcd.input_rows[0].num_columns() == n_keys
+
+        self._set_dict_keys(lcd.input_keys)
+
+        if len(self.output_buffer) != n:
+            self.output_buffer = [None]*n
+
+        for i in range(n):
+            arg_dict = self.arg_dict_base.copy()
+
+            for j in range(n_keys):
+                arg_dict[self.keys[j]] = pyobject_from_flexible_type(lcd.input_rows[0].at(i).at(j))
+
+            self.output_buffer[i] = self.lambda_function(arg_dict)
+
+        process_common_typed_list(lcd.output_values, self.output_buffer, lcd.output_enum_type)
+
+    @cython.boundscheck(False)
+    cdef process_output_dict(self, flexible_type* output_values, list mut_keys, dict ref_dict, dict ret_dict):
+        """
+        Checks to make sure all keys in ret_dict are in ref_dict, and updates
+        ret_dict with values from ref_dict as needed.  Puts the resulting values
+        in ret_dict into the position of output_values corresponding to the
+        location in mut_keys.  Furthermore, if any values are mutated in mut_keys,
+        then they are checked for equality. 
+        """
+
+        cdef long j
+        cdef long src_key_count = 0
+        cdef object src_val
+        cdef list bad_values
+
+        for j, k in enumerate(mut_keys):
+
+            try:
+                src_val = ret_dict[k]
+                src_key_count += 1
+                
+            except KeyError:                
+                src_val = ref_dict[k]
```

Comment:
It's not actually copied -- ref vs. ret.    One is the reference block, the other is the returned block from the user.